File: web/src/types/campaign.ts
export interface Campaign {
  id: string;
  name: string;
  subject?: string | null;
  body?: string | null;
  sermonId: string;
  sermonTitle?: string | null;
  recipientSource: "group" | "members";
  groupId?: string | null;
  recipientCount: number;
  status: string;
  sendAt: string;
  createdAt: string;
  updatedAt: string;
}

export interface CreateCampaignInput {
  name: string;
  subject?: string;
  body?: string;
  sermonId: string;
  groupId?: string;
  memberIds?: string[];
  sendAt: string;
}

export type CampaignUpdate = Partial<
  Omit<CreateCampaignInput, "sermonId" | "groupId" | "memberIds">
>;

File: api/app/models/campaign.py
import uuid
from datetime import datetime, time

from sqlalchemy import DateTime, ForeignKey, String, Text, Uuid, func
from sqlalchemy.orm import Mapped, mapped_column

from app.db import Base


class Campaign(Base):
    __tablename__ = "campaigns"

    id: Mapped[uuid.UUID] = mapped_column(Uuid, primary_key=True, default=uuid.uuid4)
    created_by_user_id: Mapped[uuid.UUID] = mapped_column(Uuid, nullable=False, index=True)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    subject: Mapped[str | None] = mapped_column(String(300), nullable=True)
    body: Mapped[str | None] = mapped_column(Text, nullable=True)
    recipient_source: Mapped[str] = mapped_column(String(20), nullable=False)
    group_id: Mapped[uuid.UUID | None] = mapped_column(
        Uuid, ForeignKey("groups.id", ondelete="SET NULL"), nullable=True
    )
    sermon_id: Mapped[uuid.UUID | None] = mapped_column(
        Uuid, ForeignKey("sermons.id", ondelete="SET NULL"), nullable=True, index=True
    )
    send_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True, index=True)
    weekly_day: Mapped[int | None] = mapped_column(nullable=True)
    weekly_time: Mapped[time | None] = mapped_column(nullable=True)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="draft", server_default="draft")
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now(), onupdate=func.now())


class CampaignMember(Base):
    __tablename__ = "campaign_members"

    campaign_id: Mapped[uuid.UUID] = mapped_column(
        Uuid, ForeignKey("campaigns.id", ondelete="CASCADE"), primary_key=True
    )
    member_id: Mapped[uuid.UUID] = mapped_column(
        Uuid, ForeignKey("members.id", ondelete="CASCADE"), primary_key=True, index=True
    )
